### ISSUE TITLE
Strange Multi-GPU Performance

### ISSUE THREAD

**Issue by ascenator** [tensorflow/tensorflow#1397]
I am encountering a strange Multi-GPU performance issue. To validate I've run the following script [TensorFlow Examples: Multi-GPU basics](https://github.com/aymericdamien/TensorFlow-Examples/blob/master/multigpu_basics.py).

The performance on a single GPU was 19 seconds while dual GPU computation took 28 seconds.

The same effect was reproducible on my system when I used multi-GPU for computing my own networks. You can see a short description and an ASCII illustration of my architecture here: [stackoverflow link](http://stackoverflow.com/questions/35822251/tensorflow-one-network-two-gpus).

The problem seems to get worse with bigger computations, since my own network is quite big and on a **single** GPU it is about **10 times** faster than on two GPUs :/

##### Operating System information:
`uname -or`:
> 3.10.0-327.10.1.el7.x86_64 GNU/Linux

`cat /etc/redhat-release`:
> CentOS Linux release 7.2.1511 (Core)

##### Hardware

Graphics card is a [NVIDIA Dual K80 Tesla](http://www.nvidia.com/object/tesla-k80.html). 

##### Software environment:

`pip2.7 freeze`:
> backports-abc==0.4
> backports.ssl-match-hostname==3.5.0.1
> certifi==2015.11.20.1
> cycler==0.10.0
> Cython==0.23.4
> enum34==1.1.2
> h5py==2.5.0
> libxml2-python==2.9.3
> Mako==1.0.1
> matplotlib==1.5.1
> mercurial==3.6.2
> nose==1.3.7
> numpy==1.10.4
> pkgconfig==1.1.0
> protobuf==3.0.0b2
> pyparsing==2.1.0
> python-dateutil==2.4.2
> pytz==2015.7
> singledispatch==3.4.0.3
> six==1.10.0
> tensorflow==0.7.1
> tornado==4.3
> virtualenv==13.1.2
> wheel==0.29.0

`python -c "import tensorflow; print(tensorflow.__version__)`:

> I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcublas.so.7.0 locally
> I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcudnn.so.4.0.7 locally
> I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcufft.so.7.0 locally
> I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcuda.so.1 locally
> I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcurand.so.7.0 locally
> 0.7.1

Built with **bazel** from sources, commit hash:
> commit eda89e930cfcbd992ecacafd40267d733e2153dc

### What have you tried?
1. different scripts and setups

### Logs or other output that would be helpful
Complete output of `multigpu_basics.py` with `tf.ConfigProto(log_device_placement=log_device_placement)` is attached.
[out.txt](https://github.com/tensorflow/tensorflow/files/160210/out.txt)

**Comment by thinxer** [comment]
You should try place variables on GPU. The cifar10 example uses variables on CPU by default. You may achieve about 1.5x speed up compared to a single GPU setup.

I've got about 2x speed up with 4 GPUs, and 3x with 8 GPUs. It is far from linear acceleration. I guess `tf.reduce_mean` has a noticeable overhead here, when averaging grads from different GPUs.

Even when only one GPU is used, adding the `tf.reduce_mean` operation can cause a 50% penalty on speed.

The above results are collected with Titan X GPUs and a VGG-16 network.

**Comment by ascenator** [comment]
I was not trying the CIFAR10 example, but could do this later. The author of this very simple example: [TensorFlow Examples: Multi-GPU basics](https://github.com/aymericdamien/TensorFlow-Examples/blob/master/multigpu_basics.py) claims a speedup with Multi-GPU setup while on my system it actually took longer compared with single GPU calculation.

The same thing happened when I placed some `with tf.device("/gpu:0")` and `with tf.device("/gpu:1")` here and there, where my code is able to run in parallel without creating towers.

Could you supply an example? Maybe a complete "benchmark" for Multi-GPU, to see if this error is reproducible?

**Comment by thinxer** [comment]
Ah I see. The example is not well written.

The first run is dominated by the initializing time. You should time the second run of `sess.run`.

Here's my result:

```
Single GPU computation time: 0:00:08.477408
Multi GPU computation time: 0:00:04.451062
```

**Comment by ascenator** [comment]
Aren't `a` and `b` initialized in both runs?!

My result was:
> Single GPU computation time: 0:00:19.211597
> Multi GPU computation time: 0:00:28.120005

**Comment by thinxer** [comment]
The graph is not initialized until the first run in a session. Just defining it won't do.

Your results are strange. I cannot explain it.

**Comment by ascenator** [comment]
Do I maybe have to set some memory constraints? Since I do not have two graphics cards but one dual-GPU card? Perhaps TensorFlow is doing something strange on the dual-GPU card's memory?

**Comment by johcarvajal** [comment]
I think it has to do with the Dual GPU architecture for Tesla K80. It has a PCIe switch to communicate both GPU cards internally. It shall introduce an overhead on communication. See [block diagram](https://www.microway.com/hpc-tech-tips/introducing-nvidia-tesla-k80-gpu-accelerator-kepler-gk210/)

**Comment by ascenator** [comment]
So the shared PCIe lanes are the bottleneck? Any tips on how to avoid this problem?

**Comment by johcarvajal** [comment]
I believe it has to be addressed at kernel level

**Comment by girving** [comment]
Closing since the discussion indicates it may not be a TensorFlow specific issue.  Please comment if more information appears or file a separate issue.
